### PR TITLE
Add basic documentation

### LINKED ISSUE #17: Additional information on pick_ik
Hi there,

Hate to be that guy asking for more documentation when something is actively being developed but I hope these questions won't take much time to answer.

Here are some basic questions I have:
* Can I always default to `pick_ik` as the kinematics_solver for my move group which is a kinematic chain from base link to tooltip? ie, can I replace the `kdl` plugin with `pick_ik` in my `kinematics.yaml`? Or do you recommend having a separate group that uses `pick_ik` and then using this group for only the pick and place motions?
* Do you forsee any issues with using `pick_ik` for a SCARA type robot where one of the links is prismatic? I was able to get `kdl` to work for such a robot only after setting `position_only_ik: true` in my `kinematics.yaml`. 
* How does performance compare to `kdl`? Does `pick_ik` do any caching or would the user need to implement a cached ik plugin of `pick_ik` (assuming caching does not worsen performance for this plugin)?
* Is there an example of a `kinematics.yaml` file that is configured to use `pick_ik`?

Thanks in advance!

### PR DESCRIPTION
**Pull request by sea-bass**
This PR adds basic documentation to `pick_ik`.

Closes #17.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,23 +1,20 @@
 # pick_ik
 
-## Local development in devcontainer
+`pick_ik` is an inverse kinematics (IK) solver compatible with [MoveIt 2](https://github.com/ros-planning/moveit2).
 
-0. Install docker and add yourself to the docker group.
+The solver is a reimplementation of [`bio_ik`](https://github.com/PickNikRobotics/bio_ik), which combines
+* A local optimizer which solves inverse kinematics via gradient descent
```
Comment (henningkayser) [inline README.md:6]: Please reference the specific solvers that were reimplemented. If it's something different, I would highlight that as well.

File: doc/INSTALL.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+```shell
+cd /path/to/your/workspace/src
+git clone -b main https://github.com/PickNikRobotics/pick_ik.git
+```
+
+2. Import dependencies.
+
+```shell
+sudo apt install python3-vcstool
+vcs import src < upstream.repos
```
Comment (henningkayser) [inline doc/INSTALL.md:26]:
```suggestion
vcs import src < pick_ik/upstream.repos
```

File: doc/INSTALL.md [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+```
+
+---
+
+## Install from source
+
+1. Clone this repository in a Colcon workspace.
+
+```shell
+cd /path/to/your/workspace/src
```
Comment (henningkayser) [inline doc/INSTALL.md:18]: Consider using a `COLCON_WS` variable like in the MoveIt install instructions or other tutorials

File: doc/INSTALL.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+NOTE: For now, there is no binary release for [RSL](https://github.com/PickNikRobotics/RSL), so you must add the `shared` Colcon mixin as follows:
+
+```shell
+cd /path/to/your/workspace
+colcon build --mixin release shared
+```
+
+---
+
+## Local development in devcontainer
```
Comment (henningkayser) [inline doc/INSTALL.md:49]: sweet

File: doc/USAGE.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
+# pick_ik : Usage
+
+## Using pick_ik as a Kinematics Plugin
+
+As discussed in the MoveIt2 documentation, you can use the [MoveIt Setup Assistant](https://moveit.picknik.ai/humble/doc/examples/setup_assistant/setup_assistant_tutorial.html) or change the [`kinematics.yaml` file](https://moveit.picknik.ai/main/doc/examples/kinematics_configuration/kinematics_configuration_tutorial.html?highlight=kinematics%20yaml#kinematics-configuration) for your robot setup to use `pick_ik` as the IK solver.
```
Comment (henningkayser) [inline doc/USAGE.md:5]:
```suggestion
As discussed in the MoveIt 2 documentation, you can use the [MoveIt Setup Assistant](https://moveit.picknik.ai/humble/doc/examples/setup_assistant/setup_assistant_tutorial.html) or change the [`kinematics.yaml` file](https://moveit.picknik.ai/main/doc/examples/kinematics_configuration/kinematics_configuration_tutorial.html?highlight=kinematics%20yaml#kinematics-configuration) for your robot setup to use `pick_ik` as the IK solver.
```

File: doc/USAGE.md [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+Some key parameters you may want to start with are:
+
+* `mode`: If you choose `local`, this solver will only do local gradient descent; If you choose `global`, it will also enable the evolutionary algorithm. Using the global solver will be less performant, but if you're having trouble getting out of local minima, this could help you. We recommend using `local` for things like relative motion / Cartesian interpolation / endpoint jogging, and `global` if you need to solve for goals with a far-away initial conditions.
+* `memetic_<property>`: All the properties that only kick in if you use the `global` solver. The key one is `memetic_num_threads`, as we have enabled the evolutionary algorithm to solve on multiple threads.
+* `cost_threshold`: This solver works by setting up cost functions based on how far away your pose is, how much your joints move relative to the initial guess, and custom cost functions you can add. Optimization succeeds only if the cost is less than `cost_threshold`. Note that if you're adding custom cost functions, you may want to set this threshold fairly high and rely on `twist_threshold` to be your deciding factor whereas this is more of a guideline.
```
Comment (henningkayser) [inline doc/USAGE.md:34]: Probably off-topic to this PR, but I would somewhat prefer slightly different parameter names.
* `_threshold` actually could be replaced with `*_limit` or a `max_*`
* twist implies velocity afaik, a simple `cartesian_distance_limit` or `pose_distance_limit` would fit better
* `rotation_scale` should be `rotation_weight` if it directly applies to the orientation cost
* If all of these apply to costs, adding a cost like `rotation_cost_weight` wouldn't hurt

These don't have to be addressed right now, but if we want to support possible more default cost types in the future, we should talk about how to make these parameters more consistent and maintainable in general.

File: doc/USAGE.md [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+* `rotation_scale`: If you want position-only IK, set this to 0.0. If you want to treat position and orientation equally, set this to 1.0. You can also use any value in between; it's part of the cost function. Note that the exit condition on `twist_threshold` will ignore orientation if you use `rotation_scale = 0.0`.
+* `minimal_displacement_weight`: This is one of the standard cost functions that checks for the joint angle difference between the initial guess and the solution. If you're solving for far-away goals, leave it to zero or it will hike up your cost function for no reason. Have this to a small non-zero value (e.g., 0.001) if you're doing things like Cartesian interpolation along a path, or endpoint jogging for servoing.
+
+You can test out this solver live in RViz, as this plugin is programmed to respond to parameter changes at every solve. This means that you can change values on the fly using the ROS 2 command-line interface, e.g.,
+
+```shell
+ros2 param set /rviz2 robot_description_kinematics.panda_arm.mode global
```
Comment (henningkayser) [inline doc/USAGE.md:42]: Nice example for dynamic parameter switching. Maybe we should add a "Powered by generate_parameter_library"?

File: doc/USAGE.md [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+ros2 param set /rviz2 robot_description_kinematics.panda_arm.minimal_displacement_weight 0.001
+```
+
+---
+
+## Custom Cost Functions
+
+The [kinematics plugin](../src/pick_ik_plugin.cpp) allows you to pass in an additional argument of type `IkCostFn`, which can be passed in from common entrypoints such as `RobotState::setFromIK()`. See [this page](https://moveit.picknik.ai/humble/doc/examples/robot_model_and_robot_state/robot_model_and_robot_state_tutorial.html?highlight=setfromik#inverse-kinematics) for a usage example.
+
+Alternatively, feel free to add your own cost functions to the `pick_ik` source code itself (specifically, in [`goal.hpp`](../include/goal.hpp) and [`goal.cpp`](../src/goal.cpp) and submit a pull request with the new functionality you add.
```
Comment (henningkayser) [inline doc/USAGE.md:53]:
```suggestion
Alternatively, feel free to add your own cost functions to the `pick_ik` source code yourself (specifically, in [`goal.hpp`](../include/goal.hpp) and [`goal.cpp`](../src/goal.cpp) and submit a pull request with the new functionality.
```
?

**Review by henningkayser — COMMENTED** [review]
Great documentation overall!

File: README.md (hunk 1 above)
Comment (henningkayser) [reply README.md:6]: Also, I would link the original repo https://github.com/TAMS-Group/bio_ik

File: doc/USAGE.md (hunk 6 above)
Comment (sea-bass) [reply doc/USAGE.md:34]: I've captured these in https://github.com/PickNikRobotics/pick_ik/issues/20 -- thanks!

File: doc/USAGE.md (hunk 7 above)
Comment (sea-bass) [reply doc/USAGE.md:42]: Done!

**Pull request merged**